File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
import sys
input = sys.stdin.readline

n,m = list(map(int,input().split()))
tree_list = list(map(int,input().split()))


def find_tree(tree_list):
    left = 1
    right = max(tree_list)
    while left <= right:
        mid = (left + right) // 2
        total_tree = 0
        for tree in tree_list:
            sub = tree - mid
            if sub > 0:
                total_tree += (tree - mid)
        if total_tree >= m:
            left = mid  + 1
        else:
            right = mid  - 1
        
    return right


print(find_tree(tree_list))

File: 백준/Silver/2805. 나무 자르기/나무 자르기.js
const input = require('fs').readFileSync('/dev/stdin', 'utf-8').trim().split('\n');
const [N, M] = input[0].split(' ').map(Number);
const treeList = input[1].split(' ').map(Number);

let start = 0;
let end = Math.max(...treeList);
let result = 0;

while (start <= end) {
    let mid = Math.floor((end + start) / 2);
    let totalCut = 0;

    for (let tree of treeList) {
        if (mid < tree) {
            totalCut += tree - mid;
        }
    }

    if (totalCut >= M) {
        result = mid;
        start = mid + 1;
    } else {
        end = mid - 1;
    }
}

console.log(result);
